File: java/src/main/java/cn/leetcode/multithreading/$1115_PrintFooBarAlternately.java
package cn.leetcode.multithreading;

import java.util.concurrent.Semaphore;

public class $1115_PrintFooBarAlternately {
    private int n;

    public $1115_PrintFooBarAlternately(int n) {
        this.n = n;
    }

    public void foo(Runnable printFoo) throws InterruptedException {

        for (int i = 0; i < n; i++) {
            fooSema.acquire();//值为1的时候，能拿到，执行下面的操作
            // printFoo.run() outputs "foo". Do not change or remove this line.
            printFoo.run();
            barSema.release();//释放许可给barSema这个信号量 barSema 的值+1
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {

        for (int i = 0; i < n; i++) {
            barSema.acquire();//值为1的时候，能拿到，执行下面的操作
            // printBar.run() outputs "bar". Do not change or remove this line.
            printBar.run();
            fooSema.release();//释放许可给fooSema这个信号量 fooSema 的值+1
        }
    }

    // 方法1:Semaphore
    private Semaphore fooSema = new Semaphore(1);
    private Semaphore barSema = new Semaphore(0);

}

File: python/multithreading/$1115_PrintFooBarAlternately.py
from threading import Lock


class FooBar:
    def __init__(self, n):
        self.n = n
        self.FooLock = Lock()
        self.BarLock = Lock()
        self.BarLock.acquire()

    def foo(self, printFoo: 'Callable[[], None]') -> None:

        for i in range(self.n):
            self.FooLock.acquire()
            # printFoo() outputs "foo". Do not change or remove this line.
            printFoo()
            self.BarLock.release()

    def bar(self, printBar: 'Callable[[], None]') -> None:

        for i in range(self.n):
            self.BarLock.acquire()
            # printBar() outputs "bar". Do not change or remove this line.
            printBar()
            self.FooLock.release()
